File: python/examples/evaluate_access.py
#!/usr/bin/env python3
"""Example: AuthZen evaluation — check if an agent can invoke a tool.

Usage:
    python evaluate_access.py
"""

from kswitch import KSwitchClient


def main():
    with KSwitchClient(base_url="https://localhost:5001", token="your-token", verify_ssl=False) as client:
        # Single evaluation
        print("Single AuthZen evaluation:")
        decision = client.authzen.evaluate(
            subject={"type": "agent", "id": "agent-001"},
            resource={"type": "tool", "id": "read-database"},
            action={"name": "invoke"},
            context={"environment": "production"},
        )
        print(f"  Decision: {'ALLOW' if decision.decision else 'DENY'}")
        print(f"  Context: {decision.context}")

        # Batch evaluation
        print("\nBatch AuthZen evaluation:")
        batch = client.authzen.evaluate_batch(
            evaluations=[
                {
                    "subject": {"type": "agent", "id": "agent-001"},
                    "resource": {"type": "tool", "id": "read-database"},
                    "action": {"name": "invoke"},
                },
                {
                    "subject": {"type": "agent", "id": "agent-001"},
                    "resource": {"type": "tool", "id": "write-database"},
                    "action": {"name": "invoke"},
                },
                {
                    "subject": {"type": "agent", "id": "agent-001"},
                    "resource": {"type": "tool", "id": "delete-records"},
                    "action": {"name": "invoke"},
                },
            ]
        )
        for i, eval_result in enumerate(batch.evaluations):
            print(f"  Evaluation {i + 1}: {'ALLOW' if eval_result.decision else 'DENY'}")

        # Search: what resources can this agent access?
        print("\nResource search:")
        resources = client.authzen.search_resource(
            subject={"type": "agent", "id": "agent-001"},
            action={"name": "invoke"},
        )
        for r in resources.results:
            print(f"  - {r}")

        # Search: what actions can this agent perform on a resource?
        print("\nAction search:")
        actions = client.authzen.search_action(
            subject={"type": "agent", "id": "agent-001"},
            resource={"type": "tool", "id": "read-database"},
        )
        for a in actions.results:
            print(f"  - {a}")

        # Cedar policy evaluation via policy engine
        print("\nDirect policy evaluation:")
        policy_decision = client.policy.evaluate(
            principal="agent::agent-001",
            action="invoke",
            resource="tool::read-database",
            context={"environment": "production"},
        )
        print(f"  Decision: {policy_decision.decision}")
        print(f"  Policies matched: {policy_decision.policies_matched}")

        # AuthZen discovery
        print("\nAuthZen discovery:")
        config = client.authzen.discovery()
        print(f"  Configuration: {config}")


if __name__ == "__main__":
    main()

File: typescript/examples/evaluate-access.ts
/**
 * Example: Evaluate access decisions using AuthZen and Cedar policies.
 *
 * Usage:
 *   npx tsx examples/evaluate-access.ts
 */

import { KSwitchClient, type AuthZenEvaluationRequest } from "../src/index.js";

async function main() {
  const client = new KSwitchClient({
    baseUrl: process.env.KSWITCH_BASE_URL ?? "http://localhost:5001",
    token: process.env.KSWITCH_AUTH_TOKEN,
  });

  // 1. Single AuthZen evaluation
  const request: AuthZenEvaluationRequest = {
    subject: {
      type: "agent",
      id: "agent-analytics-001",
      properties: { risk_tier: "tier_2", division: "engineering" },
    },
    resource: {
      type: "mcp_tool",
      id: "postgres-query",
      properties: { mcp_server: "mcp-postgres-readonly" },
    },
    action: {
      name: "invoke",
    },
  };

  const result = await client.authzen.evaluate(request);
  console.log("AuthZen decision:", result.decision);
  console.log("Context:", result.context);

  // 2. Batch evaluation
  const batchResult = await client.authzen.evaluateBatch({
    evaluations: [
      request,
      {
        subject: { type: "agent", id: "agent-analytics-001" },
        resource: { type: "mcp_tool", id: "s3-write" },
        action: { name: "invoke" },
      },
    ],
  });
  console.log("Batch results:", batchResult.evaluations);

  // 3. Cedar policy evaluation
  const policyDecision = await client.policy.evaluate({
    principal: "Agent::\"agent-analytics-001\"",
    action: "Action::\"invoke_tool\"",
    resource: "MCPTool::\"postgres-query\"",
    context: { environment: "production" },
  });
  console.log("Policy decision:", policyDecision.decision);

  // 4. Search for accessible resources
  const resources = await client.authzen.searchResources({
    subject: { type: "agent", id: "agent-analytics-001" },
    action: { name: "invoke" },
  });
  console.log("Accessible resources:", resources);
}

main().catch(console.error);
